Add technote covering mocked gateways for faster development (#263)
apollographql/docs: src/content/technotes/TN0017-mocked-gateway-for-client-development.mdx

---

> title: "Using Apollo Gateway mocks to increase development velocity"

Reviewer: ```suggestion
title: "Mocking graph functionality to unblock client development"
```

To mock proposed changes (such as adding a new type/field), we'd recommend using [this sample created by Apollo's Solutions Architecture team](https://github.com/apollosolutions/apollo-faker-demo), as it requires a pre-existing GraphQL API to be running. This example works by allowing for a "patched"/modified version of your subgraph to run locally with mocked data, while leveraging the remote API for all other data.

You'll need to [publish your schema to Apollo Studio](/graphos/schema/cli-registration) for use with this sample. For reasons [why to publish to Apollo Studio, see the below section](#why-use-apollo-studio-for-a-mocked-server).

To get started, run:

```graphql
mkdir mocked_gateway
export APOLLO_KEY=key_from_studio #replace with the actual key from Apollo Studio
cd mocked_gateway
touch proposed.graphql
npx github:@apollosolutions/apollo-faker-demo --graphref <yourgraph>@<variant> --remote https://yourapi.com/graphql
```

> And then modify the `proposed.graphql` file with the proposed changes. There are configuration options available leveraging the `@graphql-tools/mock` package, and can be modified as denoted in [the sample's README file, under step 5](https://github.com/apollosolutions/apollo-faker-demo#usage).

Reviewer: ```suggestion
Then, modify the `proposed.graphql` file with the proposed changes. There are configuration options available via the `@graphql-tools/mock` package, and you can set these options as documented in [the sample's README file, under step 5](https://github.com/apollosolutions/apollo-faker-demo#usage).
```

> ## Why use Apollo Studio for a mocked server?

Reviewer: ```suggestion
## Why use GraphOS for a mocked server?
```

> Now that you have either a new schema or changes, it's important to upload it into [Apollo Studio](https://studio.apollographql.com) for three reasons:

Reviewer: ```suggestion
Now that you have either a new schema or changes to an existing one, it's important to publish that schema to Apollo GraphOS for the following reasons:
```

> 1. It allows for a centralized view of the schema, along with a [mocked view of the schema](/graphos/explorer/additional-features/#mocked-responses).
> 2. It allows for the samples referenced below to automatically update on changes, allowing for a deployed version of the samples to be referenced by all developers.
> 3. For client teams using code generation tools, especially those in Apollo's clients, it's possible to use Studio as a source of the schema allowing for easier development.

Reviewer: ```suggestion
- Apollo Studio provides a centralized view of your schema, along with a [mocked view of the schema](/graphos/explorer/additional-features/#mocked-responses).
- By publishing to GraphOS, the samples referenced above can automatically update on changes, allowing for a deployed version of the samples to be referenced by all developers.
- For client teams using code generation tools (especially those in Apollo's client libraries) it's possible to use GraphOS as a source of the schema, allowing for more straightforward development.
```
